week_4/class_rational_part_1/class_rational_part_1.cpp

```diff
@@ -0,0 +1,120 @@
+#include <iostream>
+
+class Rational {
+ public:
+  Rational() {
+    p = 0;
+    q = 1;
+    divider = 1;
+  }
+
+  Rational(int numerator, int denominator) {
+    if (numerator == 0) {
+      p = numerator;
+      q = 1;
+    }
+
+    else if (numerator < 0 && denominator < 0) {
+      p = abs(numerator);
+      q = abs(denominator);
+    } else if (numerator < 0 || denominator < 0) {
+      p = abs(numerator) * (-1);
+      q = abs(denominator);
+    } else {
+      p = numerator;
+      q = denominator;
+    }
+    divider = Divider();
+  }
+
+  int Divider() {
+    if (p == q) {
+      return p;
+    } else if (p == 0 && q != 0) {
+      return q;
+    } else if (p != 0 && q == 0) {
+      return p;
+    } else {
+      int dividend_t = 0;
+      int divider_t = 0;
+      int reminder = 0;
+      if (p > q) {
+        dividend_t = p;
+        divider_t = q;
+      } else {
+        dividend_t = q;
+        divider_t = p;
+      }
+      reminder = dividend_t % divider_t;
+      while (reminder != 0) {
+        dividend_t = divider_t;
+        divider_t = reminder;
+        reminder = dividend_t % divider_t;
+      }
+      return abs(divider_t);
+    }
+  }
+
+  int Numerator() const {
+    return (p / divider);
+  }
+
+  int Denominator() const {
+    return (q / divider);
+  }
+
+ private:
+  int divider;
+  int p;
+  int q;
+};
+
+int main() {
+  {
+    const Rational r(3, 10);
+    if (r.Numerator() != 3 || r.Denominator() != 10) {
+      std::cout << "Rational(3, 10) != 3/10" << std::endl;
+      return 1;
+    }
+  }
+  {
+    const Rational r(8, 12);
+    if (r.Numerator() != 2 || r.Denominator() != 3) {
+      std::cout << "Rational(8, 12) != 2/3" << std::endl;
```

Comment:
@Ypostolov Divide this test for separate fucntions

Reply:
Please clarify the comment

Reply:
@Ypostolov Do not use extra scopes in main function create few functions for each test instead and call them sequentially from the main
...
test_1();
test_2();
...

Reply:
Fixed in 682be2784f7a4b735014298dac5a1905393ae59a
Please take a look.